File: src/stacksAndQueues/python/stackMin/min_stack.py
import sys


class minStack:
    def __init__(self):
        self.m = []
        self.stack = []

    def min(self) -> int:
        if len(self.m) == 0:
            return sys.maxsize
        return self.m[-1]

    def push(self, value) -> None:
        temp = self.min()
        if value < temp:
            self.m.append(value)
        self.stack.append(value)

    def pop(self) -> int:
        self.m.pop()
        return self.stack.pop()

    def peek(self) -> int:
        return self.stack[-1]

File: src/stacksAndQueues/java/stackMin/minStack.java
package stacksAndQueues.java.stackMin;

import java.util.Stack;

public class minStack extends Stack<Integer>{
	Stack<Integer> s;
	public minStack(){
		s = new Stack<>();
	}

	public int min(){
		if(s.isEmpty()) return Integer.MAX_VALUE;
		return s.peek();
	}

	public void push(int value){
		super.push(value);
		if(value < min()) s.push(value);
		else s.push(s.peek());
	}

	public Integer pop(){
		s.pop();
		return super.pop();
	}
}
